### PR TITLE
Flow tutorial 2021

### PR DESCRIPTION
**Pull request by marcushellberg**
Updated the "Building modern web apps with Spring Boot and Vaadin" to use Vaadin Start which helped to remove/simplify several chapters. 

Things that need to be done before merging:

- [x] https://github.com/vaadin/start/issues/1426 
- [x] Send to Doug for editing

### CODE REVIEW

File: articles/flow/tutorials/_heroku-steps.adoc [hunk 1]
```diff
@@ -0,0 +1,114 @@
+== Preparing the Application for Production
+
+It's important to build a separate production-optimized version of the application before deploying it.
```
Comment (dougjewsbury) [inline articles/flow/tutorials/_heroku-steps.adoc:3]:
```suggestion
It's important to build a separate, production-optimized version of the application before deploying it.
```

File: articles/flow/tutorials/_heroku-steps.adoc [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+In development mode, Vaadin has a live-reload widget, debug logging, and uses a quick, but unoptimized, frontend build that includes source maps for easy debugging.
+Unoptimized frontend bundles can contain several megabytes of JavaScript.
+
+The `pom.xml` build includes a `production` profile configuration that prepares an optimized build that's ready for production.
```
Comment (dougjewsbury) [inline articles/flow/tutorials/_heroku-steps.adoc:7]:
```suggestion
The `pom.xml` build includes a `production` profile configuration that prepares an optimized build which is ready for production.
```

File: articles/flow/tutorials/_heroku-steps.adoc [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+=== Using a PostgreSQL Database in Production
+
+During development, the application has used an in-memory H2 database.
```
Comment (dougjewsbury) [inline articles/flow/tutorials/_heroku-steps.adoc:11]:
```suggestion
During development, the application uses an in-memory H2 database.
```

File: articles/flow/tutorials/_heroku-steps.adoc [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+It is convenient and works well for a single user.
```
Comment (dougjewsbury) [inline articles/flow/tutorials/_heroku-steps.adoc:12]:
```suggestion
This is convenient and works well for a single user.
```

File: articles/flow/tutorials/_heroku-steps.adoc [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+In production, you want to use something more robust and persistent.
```
Comment (dougjewsbury) [inline articles/flow/tutorials/_heroku-steps.adoc:13]:
```suggestion
In production, it is better to use something more robust and persistent.
```

File: articles/flow/tutorials/_heroku-steps.adoc [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+Heroku's free tier supports PostgreSQL, so can configure your application to use that.
```
Comment (dougjewsbury) [inline articles/flow/tutorials/_heroku-steps.adoc:14]:
```suggestion
Heroku's free tier supports PostgreSQL, so you can configure your application to use that.
```

File: articles/flow/tutorials/_heroku-steps.adoc [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+Build the application with the `production` profile:
+
+[source,bash]
+----
+mvn clean package -Pproduction
+----
+
+This builds a production-optimized JAR file in the `target` folder.
+
+== Creating a Heroku Account and Installing Heroku CLI
```
Comment (dougjewsbury) [inline articles/flow/tutorials/_heroku-steps.adoc:60]:
```suggestion
== Creating a Heroku Account and Installing the Heroku CLI
```

File: articles/flow/tutorials/in-depth-course/creating-a-component.adoc [hunk 8]
```diff
@@ -21,169 +28,137 @@ The top-level view is used mainly to orchestrate the components.
...
+    save.addClickShortcut(Key.ENTER); // <6>
     close.addClickShortcut(Key.ESCAPE);
-
-    return new HorizontalLayout(save, delete, close); // <8>
+    
+    return new HorizontalLayout(save, delete, close); // <7>
   }
 }
 ----
 <1> `ContactForm` extends `FormLayout`: a responsive layout that shows form fields in 1 or 2 columns depending on viewport width.
```
Comment (dougjewsbury) [inline articles/flow/tutorials/in-depth-course/creating-a-component.adoc:101]:
```suggestion
<1> `ContactForm` extends `FormLayout`, a responsive layout that shows form fields in 1 or 2 columns depending on viewport width.
```

File: articles/flow/tutorials/in-depth-course/database-access.adoc [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+
+    public List<Status> findAllStatuses(){
+        return statusRepository.findAll();
+    }
+}
+----
+<1> The `@Service` annotation will make this a Spring-managed service that you can inject into your view.  
+<2> Use Spring constructor injection to autowire the database repositories.
+<3> Check if there is an active filter, return either all contacts or use the repository to filter based on the string. 
+<4> Service classes often include validation and other business rules before persisting data.
```
Comment (dougjewsbury) [inline articles/flow/tutorials/in-depth-course/database-access.adoc:135]:
```suggestion
<4> Service classes often include validation and other business rules before making the data persistent. 
```

File: articles/flow/tutorials/in-depth-course/forms-and-validation.adoc [hunk 10]
```diff
@@ -1,23 +1,27 @@
...
+
+== Using Vaadin Binder to Create a Form and Validate Input
 
 A form is a collection of input fields that are connected to a data model, a `Contact` in this case. 
 Forms validate user input and make it easy to get an object populated with input values from the UI. 
 
-Vaadin users the <<../../binding-data/tutorial-flow-components-binder#,`Binder`>> class to build forms. 
+Vaadin uses the <<../../binding-data/tutorial-flow-components-binder#,`Binder`>> class to build forms. 
 It binds UI fields to data object fields by name. 
 For instance, it takes a UI field named `firstName` and maps it to the `firstName` field of the data object, and the `lastName` field to the `lastName` field, and so on.
```
Comment (dougjewsbury) [inline articles/flow/tutorials/in-depth-course/forms-and-validation.adoc:26]:
```suggestion
For instance, it takes a UI field named `firstName` and maps it to the `firstName` field of the data object.
Similarly, it maps the `lastName` field to the `lastName` field, and so on. 
```

File: articles/flow/tutorials/in-depth-course/forms-and-validation.adoc [hunk 11]
```diff
@@ -28,129 +32,49 @@ This is why the field names in `Contact` and `ContactForm` are the same.
 Binder also supports an <<../../binding-data/tutorial-flow-components-binder-validation#,advanced API>> where you can configure data conversions and additional validation rules, but for this application, the simple API is sufficient.
```
Comment (dougjewsbury) [inline articles/flow/tutorials/in-depth-course/forms-and-validation.adoc:32]:
```suggestion
Binder also supports an <<../../binding-data/tutorial-flow-components-binder-validation#,advanced API>> where you can configure data conversions and additional validation rules.
However, for this application, the simple API is sufficient. 
```

File: articles/flow/tutorials/in-depth-course/forms-and-validation.adoc [hunk 12]
```diff
@@ -28,129 +32,49 @@ This is why the field names in `Contact` and `ContactForm` are the same.
...
+This means you can run the same validations both in the browser and before saving to the database, without duplicating code. 
 ====
 
-== Creating the Binder
+=== Defining Bean Validation Rules in Java
+
+You can define data validation rules as Java Bean Validation annotations on the Java class.
 
-The first step is to create a new binder field in the contact form. 
+You can see all the applied validation rules by inspecting `Contact.java`. The validations are placed above the field declarations like this:
```
Comment (dougjewsbury) [inline articles/flow/tutorials/in-depth-course/forms-and-validation.adoc:42]:
```suggestion
You can see all the applied validation rules by inspecting `Contact.java`. 
The validations are placed above the field declarations like this:
```

File: articles/flow/tutorials/in-depth-course/forms-and-validation.adoc [hunk 13]
```diff
@@ -28,129 +32,49 @@ This is why the field names in `Contact` and `ContactForm` are the same.
...
 
 Binder can use validation rules that are defined on the data object in the UI. 
-This means you can run the same validations in both the browser and before saving to the database, without duplicating code. 
+This means you can run the same validations both in the browser and before saving to the database, without duplicating code. 
 ====
 
-== Creating the Binder
+=== Defining Bean Validation Rules in Java
+
+You can define data validation rules as Java Bean Validation annotations on the Java class.
```
Comment (dougjewsbury) [inline articles/flow/tutorials/in-depth-course/forms-and-validation.adoc:40]:
```suggestion
You can define data validation rules as Java bean validation annotations on the Java class.
```

File: articles/flow/tutorials/in-depth-course/handling-view-state.adoc [hunk 14]
```diff
@@ -0,0 +1,211 @@
+---
+title: Handling View State
+order: 80
+page-title: Passing Data and Events Between Vaadin Components
```
Comment (dougjewsbury) [inline articles/flow/tutorials/in-depth-course/handling-view-state.adoc:4]:
```suggestion
page-title: "Passing data and events between Vaadin components"
```

File: articles/flow/tutorials/in-depth-course/handling-view-state.adoc [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+
+    private void updateList() {
+        grid.setItems(service.findAllContacts(filterText.getValue()));
+    }
+}
+----
+<1> The `closeEditor()` call at the end of the constructor
+* Sets the form contact to `null`, clearing out old values.
+* Hides the form.
+* Removes the `"editing"` CSS class from the view.
```
Comment (dougjewsbury) [inline articles/flow/tutorials/in-depth-course/handling-view-state.adoc:132]:
```suggestion
* sets the form contact to `null`, clearing out old values,
* hides the form,
* removes the `"editing"` CSS class from the view.
```

File: articles/flow/tutorials/in-depth-course/handling-view-state.adoc [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+private void deleteContact(ContactForm.DeleteEvent event) {
+    service.deleteContact(event.getContact());
+    updateList();
+    closeEditor();
+}
+----
+<1> The save event listener calls `saveContact()`. It:
+.. Uses `contactService` to save the contact in the event to the database.
+.. Updates the list.
+.. Closes the editor.
```
Comment (dougjewsbury) [inline articles/flow/tutorials/in-depth-course/handling-view-state.adoc:200]:
```suggestion
.. uses `contactService` to save the contact in the event to the database,
.. updates the list,
.. closes the editor.
```

File: articles/flow/tutorials/in-depth-course/handling-view-state.adoc [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+<2> The delete event listener calls `deleteContact()`. It:
+.. Uses `contactService` to delete the contact from the database.
+.. Updates the list.
+.. Closes the editor.
```
Comment (dougjewsbury) [inline articles/flow/tutorials/in-depth-course/handling-view-state.adoc:204]:
```suggestion
.. uses `contactService` to delete the contact from the database,
.. updates the list,
.. closes the editor.
```

File: articles/flow/tutorials/in-depth-course/navigation-and-layouts.adoc [hunk 18]
```diff
@@ -68,200 +80,137 @@ public class MainLayout extends AppLayout { // <2>
...
-    Span stats = new Span(contactService.count() + " contacts"); // <1>
-    stats.addClassName("contact-stats");
-    return stats;
-}
-----
-+
-<1> contactService.count() gives you the number of contacts in the database.
-It returns a `Span` with the count and a text explanation.
+    private Component getContactStats() {
+        Span stats = new Span(service.countContacts() + " contacts"); // <3>
```
Comment (dougjewsbury) [inline articles/flow/tutorials/in-depth-course/navigation-and-layouts.adoc:168]: There seem to be two note 3s?

File: articles/flow/tutorials/in-depth-course/production-deployment.adoc [hunk 19]
```diff
@@ -1,181 +1,48 @@
 ---
 title: Production Deployment
 order: 160
-page-title: Deploying a Vaadin Application on AWS Elastic Beanstalk
-description: Learn how to prepare a Vaadin app for deployment and how to deploy it on AWS
+page-title: Deploying a Vaadin Flow Application on Heroku
```
Comment (dougjewsbury) [inline articles/flow/tutorials/in-depth-course/production-deployment.adoc:4]:
```suggestion
page-title: "Deploying a Vaadin Flow application on Heroku"
```

File: articles/flow/tutorials/in-depth-course/unit-and-integration-testing.adoc [hunk 20]
```diff
@@ -212,25 +207,22 @@ Side note: you didn't need to add the annotation for normal application usage, a
...
+    CrmService service;
 
     // rest omitted
 }
 ----
 
-. Right click the package that contains both tests, and select [guilabel]#Run tests in 'com.vaadin.tutorial.crm.ui.views.list'#.
-+
+Right-click the package that contains both tests, and select [guilabel]#Run tests in 'com.example.application.views.list'#.
+You should see that both test classes run and result in 3 successful tests.
```
Comment (dougjewsbury) [inline articles/flow/tutorials/in-depth-course/unit-and-integration-testing.adoc:220]:
```suggestion
You should see that both test classes run and result in three successful tests.
```

File: articles/flow/tutorials/in-depth-course/unit-and-integration-testing.adoc [hunk 21]
```diff
@@ -241,35 +233,37 @@ You can improve the startup time by explicitly listing the needed dependencies i
...
-*** The form is bound to the correct `Contact` by ensuring the right name is visible in the field.
-
-. Rerun the tests. They should all pass.
+
+The test verifies that the form logic works by:
+
+* Asserting that the form is initially hidden.
+* Selecting the first item in the grid and verifying that:
+** The form is visible.
+** The form is bound to the correct `Contact` by ensuring the right name is visible in the field.
```
Comment (dougjewsbury) [inline articles/flow/tutorials/in-depth-course/unit-and-integration-testing.adoc:264]:
```suggestion
* asserting that the form is initially hidden,
* selecting the first item in the grid and verifying that:
** the form is visible,
** the form is bound to the correct `Contact` by ensuring the right name is visible in the field.
```

**Review by dougjewsbury — APPROVED** [review]

File: articles/flow/tutorials/in-depth-course/navigation-and-layouts.adoc (hunk 18 above)
Comment (magi42) [reply articles/flow/tutorials/in-depth-course/navigation-and-layouts.adoc]: Yes, fixed that.

**Pull request merged**